Enhance live branches API to support installing any desired version of WooCommerce

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Solaris team is running e2e and API tests on Pressable and WPCOM and wants to add more of them to the list. Currently, Solaris team is manually installing the desired version of WooCommerce before running the tests. This PR enhances the live branches API to support installing any desired version of WooCommerce using APIs.

Changes:

1. Update the `permission_callback` to avoid using WC functions so we can call these APIs without WooCommerce active.
2. Remove unused `$pr_name` parameter from the `install` method.
3. Make sure to clean up the temp/zip file after installation.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Run `cd plugins/woocommerce-beta-tester`
2. Run `pnpm run build:zip`
3. Create a fresh JN site and upload the zip file to the site.
4. Activate the `WooCommerce Beta Tester` plugin.
5. Go to `/wp-admin/site-editor.php`
6. Open browser console and run the following code:

```js
wp.apiFetch({
  path: '/wc-admin-test-helper/live-branches/install/v1',
  method: 'POST',
  data: {
    pr_name: '9.6.0-beta',
    download_url: 'https://github.com/woocommerce/woocommerce/releases/download/9.6.0-beta.1/woocommerce.zip',
    version: '9.6.0-beta'
  }
});
```

```js
wp.apiFetch({
  path: '/wc-admin-test-helper/live-branches/activate/v1',
  method: 'POST',
  data: {
    version: '9.6.0-beta'
  }
});
```

7. Confirm that WooCommerce 9.6.0-beta is installed and activated.
8. Run the following code:

```js
wp.apiFetch({
  path: '/wc-admin-test-helper/live-branches/deactivate/v1',
  method: 'GET',
});
```

9. Confirm that WooCommerce 9.6.0-beta is deactivated.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Enhance live branches API with permission checks and refactor installation logic

- Added a new function `check_live_branches_permissions` to validate user permissions for live branches actions.
- Updated REST route registration to use the new permission callback.
- Refactored the `install` method in `WC_Beta_Tester_Live_Branches_Installer` to remove the unnecessary PR name parameter and improve error handling.
- Adjusted CLI command to align with the updated installation method.
- Improved temporary file cleanup during installation process.

chihsuan committed:
Add changelog

ilyasfoo commented on plugins/woocommerce-beta-tester/changelog/update-beta-tester-live-branch-install:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
+
+Enhance live branches API with permission checks to allow install desired WC version without wWC exists
```
```suggestion
Enhance live branches API with permission checks to allow install desired WC version without WC exists
```

ilyasfoo commented on plugins/woocommerce-beta-tester/api/live-branches/install.php:
```diff
@@ -9,19 +9,39 @@
+ * @return bool|WP_Error
+ */
+function check_live_branches_permissions() {
+	if ( ! current_user_can( 'install_plugins' ) ) {
```
Nice improvement to use capability check instead of a role check!

ilyasfoo reviewed (approved):
Tests well, LGTM! 🚢

ilyasfoo commented:
Not related to the PR: It seems install, activate, and deactivate APIs return JSON encoded in string:

<img width="636" alt="image" src="https://github.com/user-attachments/assets/2fbc306a-9e9b-4586-bf47-f4ea19e87713" />

chihsuan committed:
Update update-beta-tester-live-branch-install

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

chihsuan commented:
> Not related to the PR: It seems install, activate, and deactivate APIs return JSON encoded in string. I'm wondering if this is intended 🤔
> 

Thanks for bringing this up! Looks like the client side doesn't use the response from the APIs so I'm not sure if this is intended. 🤔  I'm not sure if there are any side effects if we change it, so I'll leave it as is for now

rjchow reviewed (approved):
nice work, thank you!

Stojdza commented:
Thanks for working on this, @chihsuan 🙌 I tested it on behalf of Solaris, and it works as described! I successfully installed and activated WooCommerce on JN and Pressable websites. 

I noticed that when some previous version of WC is installed on the website, I cannot activate WC this way. So, I needed to deactivate WC first (step 8) and then do step 6 (install and activate). We can first deactivate WC when automated tests start executing and then install and activate the desired version, but I wonder if there is an option to `force`  install and activate WC? Here is the screenshot:

<img width="1800" alt="Screenshot 2025-01-02 at 22 34 45" src="https://github.com/user-attachments/assets/a504d052-d201-421e-9d32-d2a53f53ef67" />


Additionally, is there an option to install the `latest` version this way? So, when a new Beta or RC is out, tests start executing and pull the latest version without specifying the version, eg. `9.6.0-beta`?

chihsuan committed:
Enhance WooCommerce Beta Tester API with new parameters and force activation option

- Added new required arguments for the POST request to install a version: `download_url`, `pr_name`, and `version`.
- Introduced an optional `force` parameter for the POST request to activate a version, allowing deactivation of WooCommerce if already active.
- Updated the description for the GET request to clarify its purpose of deactivating the active WooCommerce plugin.

These changes improve the API's functionality and provide better control over plugin installation and activation processes.

chihsuan committed:
Removing the optional 'force' parameter from the activation request

chihsuan committed:
Enhance WooCommerce Beta Tester API with new installation endpoint

- Added a new REST route to install the latest WooCommerce version, including an option to include pre-releases.
- Implemented a function to fetch the latest WooCommerce release from GitHub, handling both regular and pre-release versions.
- Improved error handling for release fetching and installation processes.

These changes enhance the API's capabilities for managing WooCommerce installations.

chihsuan commented:
@Stojdza Thanks for testing it out!

> We can first deactivate WC when automated tests start executing and then install and activate the desired version, but I wonder if there is an option to `force` install and activate WC?

Good question! I tried to add a `force` option to the API, but unfortunately, it didn't work. 

When we deactivate and then reactivate the WooCommerce plugin within the same API request, the issue arises because the WooCommerce constants (WC_*) remain defined in memory (PHP constants cannot be redefined or unset once they are set).

> Additionally, is there an option to install the `latest` version this way? So, when a new Beta or RC is out, tests start executing and pull the latest version without specifying the version, eg. `9.6.0-beta`?

Yeah, I've added a new endpoint to install the latest version. It will install the latest release. You can specify the `include_pre_releases` option to include Beta/RC versions.

Please try it out and let me know if it works for you!

```js
wp.apiFetch({
  path: '/wc-admin-test-helper/live-branches/install/latest/v1',
  method: 'POST',
  data: { include_pre_releases: true }
});
```

chihsuan committed:
Merge branch 'trunk' into update/beta-tester-live-branch-install

Stojdza commented:
> When we deactivate and then reactivate the WooCommerce plugin within the same API request, the issue arises because the WooCommerce constants (WC_*) remain defined in memory (PHP constants cannot be redefined or unset once they are set).

👍 

> Yeah, I've added a new endpoint to install the latest version. It will install the latest release. You can specify the include_pre_releases option to include Beta/RC versions.

Thanks for adding this option! It works as expected. I can see `version: "9.6.0-beta.1"` in the response ✅ 

> 5. Go to /wp-admin/site-editor.php

Pardon my ignorance here - why do we need to go to this page to execute the command?

chihsuan commented:
> Thanks for adding this option! It works as expected. I can see version: "9.6.0-beta.1" in the response ✅

🙌 

> Pardon my ignorance here - why do we need to go to this page to execute the command?

We just need to go to the page where the `wp.apiFetch` object is available. 🙂  The site-editor page loads the `@wordpress/api-fetch` script. We can also use `window.fetch` directly to make the request but will need to handle the nonce and other headers manually.
